tests/SpecTests/Operation.php

```diff
@@ -500,6 +504,35 @@ private function executeForSession(Session $session, Context $context)
         }
     }
 
+    private function executeForTestRunner(FunctionalTestCase $test, Context $context)
+    {
+        switch ($this->name) {
+            case 'assertSessionPinned':
+                $session = $context->{$this->arguments['session']};
```

Comment:
Rather than dynamically access the Context property, I think this method should start by calling `replaceArgumentSessionPlaceholder()` as is done in other `execute*()` methods within this class.

That will replace the placeholder value (e.g. "session0") with the appropriate session object and also raise a meaningful exception if the placeholder is invalid. After calling that method, we can trust that the argument is a Session instance.

Should this have been picked up by CS, or do the sniffs ignore dynamic attribute access?

Reply:
phpcs can't inspect this as it has no knowledge of types. This would be caught by using static analysis with phpstan or psalm at a stricter level.